addons/sale/models/sale.py

```diff
@@ -1289,7 +1306,23 @@ def _onchange_discount(self):
             if discount > 0:
                 self.discount = discount
 
-
     def _is_delivery(self):
         self.ensure_one()
-        return False
\ No newline at end of file
+        return False
+
+    _sql_constraints = [
+        ('accountable_product_id_required', "CHECK(line_type != 'line_accountable' OR product_id IS NOT NULL)", "You must specify a product."),
+        ('accountable_price_unit_required', "CHECK(line_type != 'line_accountable' OR price_unit IS NOT NULL)", "You must specify a unit price."),
+        ('accountable_product_uom_qty_required', "CHECK(line_type != 'line_accountable' OR product_uom_qty IS NOT NULL)", "You must specify a quantity."),
+        ('accountable_product_uom_required', "CHECK(line_type != 'line_accountable' OR product_uom IS NOT NULL)", "You must specify a unit of measure."),
+        ('accountable_customer_lead_required', "CHECK(line_type != 'line_accountable' OR customer_lead IS NOT NULL)", "You must specify a delivery lead time."),
+
+        ('non_section_line_pagebreak_null', "CHECK(line_type = 'line_section' OR line_pagebreak IS NULL OR line_pagebreak = FALSE)", "You cannot specify a pagebreak."),
+        ('non_section_line_subtotal_null', "CHECK(line_type = 'line_section' OR line_subtotal IS NULL OR line_subtotal = FALSE)", "You cannot specify a subtotal."),
+
+        ('non_accountable_product_id_null', "CHECK(line_type = 'line_accountable' OR product_id IS NULL)", "You cannot specify a product."),
+        ('non-accountable_price_unit_null', "CHECK(line_type = 'line_accountable' OR price_unit IS NULL OR price_unit = 0)", "You cannot specify a unit price."),
```

Comment:
Don't use dash in SQL constraint name. 
And do more explicit message: "you can not do that since ...."

Reply:
We decided with ODO to merge most of those constraints.

Since they are only there to be failsafe and not supposed to happen in a normal flow, we will reduce the error messages to minimum.